Added cleanup to server thread join method

kpayson64 opened the pull request:
Addresses #6680

My analysis at this point is that on the cloud dev environments, the servers are referenced by a top level object, and so their delete methods do not get called until interpreter exit, the issue is that there is a special thread shutdown function that gets called before any object deletion or exit handlers that attempts to join() our main server thread.  The simple fix is to make Server extend thread, and have a custom join method that does the necessary shutdown.

An easy way to reproduce on any environment is the following script

```
import unittest
from tests.unit import _rpc_test
suite = unittest.TestLoader().loadTestsFromTestCase(_rpc_test.RPCTest)
unittest.TextTestRunner(verbosity=2).run(suite)
```

Note that with the fix in this PR, this properly exits.

nathanielmanistaatgoogle commented:
Yeah, this strikes me as every bit as radical and strange as I expected. Can you link me to this top-level object that references all servers?

kpayson64 commented:
Turns out no top level reference is created, but rather a reference to a circular reference in this line
https://github.com/grpc/grpc/blob/master/src/python/grpcio/tests/_runner.py#L197
```unittest.TestCase.run(result)``` saves a reference to result, and result has references to all the test cases.  By removing the argument in the aforementioned line (as well as commenting out https://github.com/grpc/grpc/blob/master/src/python/grpcio/tests/_runner.py#L208), the test no longer hangs.

The circular reference gets cleaned up as part of garbage collection, but it looks like on the GCE instances that happens after the thread cleanup.  If I force garbage collection by adding 
```
   result = None
    loader = None
    runner = None
    import gc; gc.collect() # Without this line, it hangs
```
To the end of the run method at
https://github.com/grpc/grpc/blob/master/src/python/grpcio/commands.py#L297, the test ends cleanly.

This means that anyone with a circular reference that references a ```Server``` (but not part of the cycle)  cannot expect good cleanup behavior.

nathanielmanistaatgoogle commented:
Thank you for the investigation; these are much, much happier circumstances.

To take the last part first: "This means that anyone with a circular reference that references a Server (but not part of the cycle) cannot expect good cleanup behavior.". Yes. This is not a concern. If someone is doing something that references a running server from an object cycle, they definitely aren't going to get the fancy "server is nicely shut down when garbage collected" policy we've implemented. They'll have to shut their server down with its `stop` method. This is fine.

(In complete fairness, that answer applies equally well to us and our test situation.)

As for the rest: I don't know enough about `_runner.py` and `commands.py`, but I would consider circular references involving test case instances to be a bug in them and look for a way to fix them there rather than alter the code of our `TestCase` subclasses or (yuck!) the production code covered by the tests.

Object cycles aren't necessarily a bug, especially during the useful lives of the objects, but object cycles reaching garbage collection smells like an invitation of trouble. Let's talk in the morning?

kpayson64 commented:
Changed to add join() cleanup to server thread instead of adding thread semantics to server.  Back to @nathanielmanistaatgoogle

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,27 @@ def serialize(message, serializer):
+    self._join_kwargs = join_kwargs
+
+  def join(self, timeout=None):
+    if self._join_target:
```
gRPC Python style is to always use `is None` and `is not None` when comparing against the `None` object.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,27 @@ def serialize(message, serializer):
+  
+  def __init__(self, group=None, target=None, join_target=None, name=None,
+               args=(), join_args=(), kwargs={}, join_kwargs={}):
+    """join_target() will be called with a timeout,
```
Nearly without exception the doc string of an `__init__` method should be
```
Constructor.

Attributes:
  <attribute names and descriptions>
```
. No work of any consequence should be done in a constructor, and thus there should be nothing to say about it except to describe its parameters. See [the style guide](https://google.github.io/styleguide/pyguide.html) for [examples](https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments).

nathanielmanistaatgoogle replied:
Actually those examples in the style guide don't go far enough to illustrate what I'm saying. :-P

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,27 @@ def serialize(message, serializer):
+  we acomplish this by wrapping the join() method.
+  """
+  
+  def __init__(self, group=None, target=None, join_target=None, name=None,
```
The behavior to be called at join-time should be a required (and non-`None`) construction parameter because if it's ever `None` this class has no point in being used at that particular use-site.

Also I recommend just taking a single no-arg parameter (call it "behavior"?) and leave it to use-sites to wrap their args and kwargs in a `lambda` or custom function.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,27 @@ def serialize(message, serializer):
+
+
+class CleanupThread(threading.Thread):
+  """A wrapper around thread for custom join behavior.
```
"wrapper around" sounds too much like delegation; what we're doing here is inheritance. How about "A threading.Thread subclass supporting custom behavior at join-time."?

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,27 @@ def serialize(message, serializer):
+
+  def join(self, timeout=None):
+    if self._join_target:
+      self._join_target(timeout, *self._join_args, **self._join_kwargs)
```
A thread may be `join`ed multiple times. Must the `_join_target` be idempotent, or should this class take additional measures to ensure that the `_join_target` is only called once?

kpayson64 replied:
I've added an idempotent requirement to the doc.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,27 @@ def serialize(message, serializer):
+class CleanupThread(threading.Thread):
+  """A wrapper around thread for custom join behavior.
+  
+  On Interpreter exit, Python will attempt to join outstanding threads
```
"Interpreter" should only get a capital i when the word "Python" immediately precedes it.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_server.py:
```diff
@@ -719,6 +708,26 @@ def cancel_all_calls_after_grace():
 
+def _join_server_thread(timeout, state):
+  if not timeout:
+    _stop(state, 2.0).wait()
```
Magic number. To best emulate "wait as long as it takes", shouldn't the value passed here be astronomically high (and, sadly, an equally magic number)?

nathanielmanistaatgoogle commented:
We've learned so much in the last twenty-four hours; please amend the commit message to reflect all of it.

kpayson64 commented:
Addressed comments.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,38 @@ def serialize(message, serializer):
+      behavior (function): Function called on join() with a single
+          argument, timeout, indicating the maximum duration of
+          `behavior`, or None indicating `behavior` has no deadline.
+          `behavior` must be idempotent and not None.
```
Drop "and not None"; the rest of the doc string has already laid out many requirements that `None` fails to meet.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,38 @@ def serialize(message, serializer):
+
+  On Python Interpreter exit, Python will attempt to join outstanding threads
+  prior to garbage collection.  We may need to do additional cleanup, and
+  we acomplish this by overriding the join() method.
```
Typo "acomplish".

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_server.py:
```diff
@@ -719,6 +708,24 @@ def cancel_all_calls_after_grace():
+    state.stage = _ServerStage.STARTED
+    _request_call(state)    
+    def cleanup_server(timeout):
+      if not timeout:
```
Always favor
```
if <condition>:
  <other statements>
else:
  <some statements>
```
over
```
if not <condition>:
  <some statements>
else:
  <other statements>
```
. Except here I don't think it's a boolean comparison that best fits the situation: `timeout` may be a numeric value or it may be `None`, and I think the branch works best when aligned that way.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_common.py:
```diff
@@ -97,3 +98,38 @@ def serialize(message, serializer):
+                                        name=name, args=args, kwargs=kwargs)
+    self._behavior = behavior
+
+  def join(self, timeout=None):
```
If a caller passes a `timeout` value of 6, is it possible for this method to take twelve seconds to terminate?

kpayson64 commented:
Addressed comments and added a unit test for CleanupThread.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/_thread_cleanup_test.py:
```diff
@@ -0,0 +1,117 @@
+      self.assertEqual('arg3', arg3)
+      event.set()
+
+    t = _common.CleanupThread(behavior=lambda x: None, target=target,
```
[Single-letter identifiers are disallowed by the style guide](https://google.github.io/styleguide/pyguide.html?showone=Naming#Naming).

nathanielmanistaatgoogle commented:
Looks nearly good; just the one nit about the thread names in the test.

kpayson64 commented:
Fixed the thread name.  I also reverted the server thread cleanup to use a small default timeout instead of ```float('inf')```, the reason being that any unfinished client call could hang the server indefinitely, which defeats the purpose of this change.

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/_thread_cleanup_test.py:
```diff
@@ -0,0 +1,117 @@
+    self.assertTrue(event.is_set())
+
+  def testJoinNoTimeout(self):
+    t = _common.CleanupThread(behavior=cleanup)
```
The prohibition on single-letter identifiers applies throughout the codebase.

kpayson64 replied:
Oops...

nathanielmanistaatgoogle commented on src/python/grpcio/tests/unit/_thread_cleanup_test.py:
```diff
@@ -0,0 +1,117 @@
+    event.set()
+
+  def testJoinZeroTimeout(self):
+    t = _common.CleanupThread(behavior=cleanup)
```
... all the way down to here.

nathanielmanistaatgoogle commented:
Luhguhtum.

tseaver commented:
@nathanielmanistaatgoogle AFAICT, this fix is not yet released: is that correct? I am definitely seeing hang-on-exit behavior with grpcio-0.14.0.

nathanielmanistaatgoogle commented:
@tseaver: you are correct; this fix will be included in 0.15 and later.

tseaver commented:
@nathanielmanistaatgoogle is there an ETA for that release?

nathanielmanistaatgoogle commented:
O(14 days), we hope.

tseaver commented:
@nathanielmanistaatgoogle Thanks!
